src/main/groovy/io/openliberty/tools/gradle/tasks/DevTask.groovy

```diff
@@ -564,17 +580,21 @@ class DevTask extends AbstractFeatureTask {
                 // - start server
                 util.restartServer();
                 return true;
-            } else if (installFeatures) {
-                if (generateFeatures) {
-                    // Increment generate features on build dependency change
-                    ProjectConnection gradleConnection = initGradleProjectConnection();
-                    BuildLauncher gradleBuildLauncher = gradleConnection.newBuild();
-                    runGradleTask(gradleBuildLauncher, 'compileJava', 'processResources'); // ensure class files exist
-                    Collection<String> javaSourceClassPaths = getJavaSourceClassPaths();
-                    libertyGenerateFeatures(javaSourceClassPaths, false);
-                    libertyCreate(); // need to run create in order to copy generated config file to target
-                }
+            } else if (installFeatures && !compileProject) {
                 libertyInstallFeature();
+            } else if (compileProject && generateFeatures) { // generate features if compile dependencies have been modified
+                ProjectConnection gradleConnection = initGradleProjectConnection();
+                BuildLauncher gradleBuildLauncher = gradleConnection.newBuild();
+                runGradleTask(gradleBuildLauncher, 'compileJava', 'processResources'); // ensure class files exist
+
+                // optimize generate features on build dependency change
+                Collection<String> javaSourceClassPaths = getJavaSourceClassPaths();
+                boolean generateFeaturesSuccess = libertyGenerateFeatures(javaSourceClassPaths, true);
```

Comment:
Just realized we do not need to pass the `javaSourceClassPaths` in this case since we are optimizing. Can pass `null`.
This would also apply to https://github.com/OpenLiberty/ci.maven/blob/c4038512547184e25699d16492e6db05976fe307/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java#L915